server/routerlicious/packages/services-client/src/rollingHash.ts

```diff
@@ -0,0 +1,37 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ISequencedDocumentMessage, ISequencedDocumentSystemMessage } from "@fluidframework/protocol-definitions";
+import { str } from "crc-32";
+
+export function getNextHash(message: ISequencedDocumentMessage, lastHash: string): string {
+    const messageWithData = message as ISequencedDocumentSystemMessage;
+
+    const clientIdProperty: keyof ISequencedDocumentSystemMessage = "clientId";
+    const sequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "sequenceNumber";
+    const minimumSequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "minimumSequenceNumber";
+    const clientSequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "clientSequenceNumber";
+    const referenceSequenceNumberProperty: keyof ISequencedDocumentSystemMessage = "referenceSequenceNumber";
+    const typeProperty: keyof ISequencedDocumentSystemMessage = "type";
+    const timestampProperty: keyof ISequencedDocumentSystemMessage = "timestamp";
+    const dataProperty: keyof ISequencedDocumentSystemMessage = "data";
+
+    // Ordering of these fields decides the ordering of serialized message.
+    const fields: string[] = [
+        clientIdProperty,
+        sequenceNumberProperty,
+        minimumSequenceNumberProperty,
+        clientSequenceNumberProperty,
+        referenceSequenceNumberProperty,
+        typeProperty,
+        timestampProperty,
+        dataProperty,
+    ];
+
+    // Should we just use stringified JSON or come up with our own serialization of values?
+    const serializedMessage = JSON.stringify(messageWithData, fields);
```

Comment:
The [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify) say that `Properties of non-array objects are not guaranteed to be stringified in any particular order. Do not rely on ordering of properties within the same object within the stringification.`

If the order is not guaranteed, then maybe the hash could change even though it's the same op?

Reply:
Based on my understanding, the fields will determine the order (https://stackoverflow.com/questions/16167581/sort-object-properties-and-json-stringify). And our fields are primitive only which means we can use them easily.